### ISSUE TITLE
DnsServerAddress: Test-TargetResource always returns $false

### ISSUE THREAD

**Issue by mikekl62** [dsccommunity/NetworkingDsc#446]
#### Details of the scenario you tried and the problem that is occurring

I have a configuration that contains the DnsServerAddress resource, where I specify that DNS server addresses are managed by DHCP by excluding the Address parameter.

When I test compliancy using Test-DscConfiguration the command always reports $false, even when I run it directly after successful initial configurations. Get-DscConfigurationStatus reports Success for all subsequent consistency checks.

#### Verbose logs showing the problem

This is an extract from Test-DscConfiguration showing the output from the DnsServerAddress resource.

VERBOSE: [xxx]: LCM:  [ Start  Resource ]  [[DnsServerAddress]DnsDhcpEnable]
VERBOSE: [xxx]: LCM:  [ Start  Test     ]  [[DnsServerAddress]DnsDhcpEnable]
VERBOSE: [xxx]:                            [[DnsServerAddress]DnsDhcpEnable] Test-TargetResource: Checking the DNS server addresses.
VERBOSE: [xxx]:                            [[DnsServerAddress]DnsDhcpEnable] Get-DnsClientServerStaticAddress: Getting staticly assigned DNS server IPv4 address for interface alias 'MiddleEarth'.
VERBOSE: [xxx]:                            [[DnsServerAddress]DnsDhcpEnable] Get-DnsClientServerStaticAddress: Statically assigned DNS server IPv4 address for interface alias 'MiddleEarth' is not set.
**VERBOSE: [xxx]:                            [[DnsServerAddress]DnsDhcpEnable] Test-TargetResource: DNS server addresses are not correct. Expected "", actual "".**
VERBOSE: [xxx]: LCM:  [ End    Test     ]  [[DnsServerAddress]DnsDhcpEnable] False in 0.0930 seconds.
VERBOSE: [xxx]: LCM:  [ End    Resource ]  [[DnsServerAddress]DnsDhcpEnable]

When I saw 'Excpected "", actual ""' I got suspicioius.

#### Suggested solution to the issue

I have plowed through the source code for DnsServerAddress here on GitHub, and I suspect that the root cause is that the helper function Get-DnsClientServerStaticAddress returns $null when DNS server addresses are set by DHCP.

The return value from Get-DnsClientServerStaticAddress is assigned to $currentAddress in line 251-253 (in version 7.4.0.0).

```powershell
[String[]] $currentAddress = @(Get-DnsClientServerStaticAddress `
        @PSBoundParameters `
        -ErrorAction Stop)
```

If Get-DnsClientServerStaticAddress returns $null, this should be equivalent (unless I have missed something):

```powershell
[String[]] $currentAddress = @($null)
```

By executing the code above and then inspecting $currentAddress I draw the conclusion that $currentAddress is an array of string with one element that contains an empty string.

```powershell
PS > $currentAddress.Length
1
PS > $currentAddress[0] -eq ""
True
```

Since Address has been excluded in this case, I assume that $Address in Test-TargetResouce will be set to an empty array (line 231-244)

```powershell
if ($PSBoundParameters.ContainsKey('Address'))
{
    foreach ($ServerAddress in $Address)
    {
        Assert-ResourceProperty `
            -Address $ServerAddress `
            -AddressFamily $AddressFamily `
            -InterfaceAlias $InterfaceAlias
    } # foreach
}
else
{
    [String[]] $Address = @()
} # if
```

If my analysis is correct an empty array will be compared with an array with one element that is an empty string in line 256-259.

```powershell
[Boolean] $addressDifferent = (@(Compare-Object `
    -ReferenceObject $currentAddress `
    -DifferenceObject $Address `
    -SyncWindow 0).Length -gt 0)
```

And that should obvisionsly result in that $addressDifferent is set to $true, and that Test-TargetResource always returns $false.

Is it possible to change Get-DnsClientServerStaticAddress so @() is returned instead of $null when DNS server addresses are set by DHCP? Maybe this will break other resources, but if not it seems to be a solution.

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)

My configuration is a partial configuration for servers with IP addresses managed by DHCP.

```powershell
Configuration NetDynIP
{
    Import-DscResource -ModuleName NetworkingDsc -ModuleVersion 7.4.0.0

    Node $AllNodes.Where{$_.NodeName -eq "$NodeName"}.NodeName
    {
        NetAdapterName MiddleEarthNetAdapter
        {
            MacAddress     = "$($Node.MACAddress)"
            NewName        = "MiddleEarth"
        }

        NetIPInterface DhcpEnable
        {
            InterfaceAlias = "MiddleEarth"
            AddressFamily  = "IPv4"
            Dhcp           = "Enabled"
            DependsOn      = "[NetAdapterName]MiddleEarthNetAdapter"
        }

        DnsServerAddress DnsDhcpEnable
        {
            InterfaceAlias = "MiddleEarth"
            AddressFamily  = "IPv4"
            DependsOn      = "[NetAdapterName]MiddleEarthNetAdapter"
        }
    }
}
```

#### The operating system the target node is running

OsName               : Microsoft Windows Server 2019 Standard
OsOperatingSystemSKU : StandardServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 1809
WindowsBuildLabEx    : 17763.1.amd64fre.rs5_release.180914-1434
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### Version and build of PowerShell the target node is running

Name                           Value
----                           -----
PSVersion                      5.1.17763.1007
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17763.1007
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

#### Version of the DSC module that was used ('dev' if using current dev branch)

7.4.0.0

**Comment by phbits** [comment]
## TLDR; ##
Either comment out `Address` or leave it out entirely.

## Testing ##
I ran into this same issue while testing and found the answer in the file `NetworkingDsc/source/DSCResources/DSC_DnsServerAddress/DSC_DnsServerAddress.schema.mof` where the pertinent information is:

> [Write, Description("The desired DNS Server address(es). Exclude to enable DHCP.")] string Address[]; 

I commented out `Address`, as shown in the following DSC config, to exclude it. You'll see I had previously tried the empty string approach which didn't work.

```
Configuration DnsServerTest
{
    Import-DscResource -Module NetworkingDsc

    Node localhost
    {
        NetIPInterface Ethernet_2_IPv4_Dhcp
        {
            InterfaceAlias = 'Ethernet 2'
            AddressFamily  = 'IPv4'
            Dhcp           = 'Enabled'
        }
        DnsServerAddress Ethernet_2_IPv4_DnsServerAddress
        {
            #Address        = ''
            InterfaceAlias = 'Ethernet 2'
            AddressFamily  = 'IPv4'
        }
    }
}

. DnsServerTest -OutputPath C:\DscTest2\DnsServerTest -Verbose
```
I reset LCM as I'm been doing much testing.
```
PS C:\DscTest2> Reset-LcmConfiguration
```
Configured LCM.
```
PS C:\DscTest2> Set-LcmSetting -RebootNodeIfNeeded $false -ConfigurationMode ApplyAndAutoCorrect -ActionAfterReboot ContinueConfiguration
```
Created MOF.
```
PS C:\DscTest2> & .\DnsServerTest.ps1


    Directory: C:\DscTest2\DnsServerTest


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       10/22/2020   8:50 AM           2754 localhost.mof
```
Added configuration.
```
PS C:\DscTest2> Add-LcmPartialConfiguration -PartialName 'DnsServerTest' -Description 'DnsServerTest Partial Configuration.' -RefreshMode 'Push' -Verbose
VERBOSE: Perform operation 'Invoke CimMethod' with following parameters, ''methodName' = SendMetaConfigurationApply,'className' = MSFT_DSCLocalConfigurationManager,'namespaceName' =
root/Microsoft/Windows/DesiredStateConfiguration'.
VERBOSE: An LCM method call arrived from computer INTERNETVM2 with user sid S-1-5-21-2186093478-4284278558-843089667-500.
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Resource ]  [MSFT_DSCMetaConfiguration]
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Set      ]  [MSFT_DSCMetaConfiguration]
VERBOSE: [INTERNETVM2]: LCM:  [ End    Set      ]  [MSFT_DSCMetaConfiguration]  in 0.0270 seconds.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Resource ]  [MSFT_DSCMetaConfiguration]
VERBOSE: [INTERNETVM2]: LCM:  [ End    Set      ]
VERBOSE: [INTERNETVM2]: LCM:  [ End    Set      ]    in  0.0860 seconds.
VERBOSE: Operation 'Invoke CimMethod' complete.
VERBOSE: Set-DscLocalConfigurationManager finished in 0.475 seconds.
```
Published the configuration.
```
PS C:\DscTest2> Publish-DscConfiguration -Path C:\DscTest2\DnsServerTest\ -Verbose
VERBOSE: Perform operation 'Invoke CimMethod' with following parameters, ''methodName' = SendConfiguration,'className' = MSFT_DSCLocalConfigurationManager,'namespaceName' =
root/Microsoft/Windows/DesiredStateConfiguration'.
VERBOSE: An LCM method call arrived from computer INTERNETVM2 with user sid S-1-5-21-2186093478-4284278558-843089667-500.
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Set      ]
VERBOSE: [INTERNETVM2]: LCM:  [ End    Set      ]     Saved configuration document into the partial configuration store.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Set      ]
VERBOSE: Operation 'Invoke CimMethod' complete.
VERBOSE: Publish-DscConfiguration finished in 0.137 seconds.
```
Launched DSC.
```
PS C:\DscTest2> Start-DscConfiguration -UseExisting -Verbose -Wait
VERBOSE: Perform operation 'Invoke CimMethod' with following parameters, ''methodName' = ApplyConfiguration,'className' = MSFT_DSCLocalConfigurationManager,'namespaceName' =
root/Microsoft/Windows/DesiredStateConfiguration'.
VERBOSE: An LCM method call arrived from computer INTERNETVM2 with user sid S-1-5-21-2186093478-4284278558-843089667-500.
VERBOSE: [INTERNETVM2]:                            [] Starting consistency engine.
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Resource ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Test     ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]
VERBOSE: [INTERNETVM2]:                            [[NetIPInterface]Ethernet_2_IPv4_Dhcp] Test-TargetResource: Checking settings on net IP interface for alias 'Ethernet 2' and address
family 'IPv4'.
VERBOSE: [INTERNETVM2]:                            [[NetIPInterface]Ethernet_2_IPv4_Dhcp] Get-NetworkIPInterface: Getting settings for net IP interface for alias 'Ethernet 2' and address
family 'IPv4'.
VERBOSE: [INTERNETVM2]:                            [[NetIPInterface]Ethernet_2_IPv4_Dhcp] Test-TargetResource: The property 'Dhcp' value 'Disabled' does not match the desired value
'Enabled'.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Test     ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]  in 1.0030 seconds.
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Set      ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]
VERBOSE: [INTERNETVM2]:                            [[NetIPInterface]Ethernet_2_IPv4_Dhcp] Set-TargetResource: Applying settings to net IP interface for alias 'Ethernet 2' and address family
 'IPv4'.
VERBOSE: [INTERNETVM2]:                            [[NetIPInterface]Ethernet_2_IPv4_Dhcp] Get-NetworkIPInterface: Getting settings for net IP interface for alias 'Ethernet 2' and address
family 'IPv4'.
VERBOSE: [INTERNETVM2]:                            [[NetIPInterface]Ethernet_2_IPv4_Dhcp] Set-TargetResource: Setting Net IP interface for alias 'Ethernet 2' and address family 'IPv4'
parameter 'Dhcp' to 'Enabled'.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Set      ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]  in 0.2650 seconds.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Resource ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Resource ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Test     ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Test-TargetResource: Checking the DNS server addresses.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Get-DnsClientServerStaticAddress: Getting staticly assigned DNS server IPv4 address
for interface alias 'Ethernet 2'.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Get-DnsClientServerStaticAddress: Statically assigned DNS server IPv4 address for
interface alias 'Ethernet 2' is not set.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Test-TargetResource: DNS server addresses are not correct. Expected "", actual "".
VERBOSE: [INTERNETVM2]: LCM:  [ End    Test     ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]  in 1.3490 seconds.
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Set      ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Set-TargetResource: Applying the DNS server addresses.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Get-DnsClientServerStaticAddress: Getting staticly assigned DNS server IPv4 address
for interface alias 'Ethernet 2'.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Get-DnsClientServerStaticAddress: Statically assigned DNS server IPv4 address for
interface alias 'Ethernet 2' is not set.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Set-TargetResource: DNS server addresses were set to the desired state of DHCP.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Set      ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]  in 0.9200 seconds.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Resource ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]
VERBOSE: [INTERNETVM2]:                            [] Consistency check completed.
VERBOSE: Operation 'Invoke CimMethod' complete.
VERBOSE: Time taken for configuration job to complete is 4.57 seconds
```
The verbose messages show the interface is being properly configured to get DNS servers from DHCP. I manually checked the UI to confirm this and proceeded to manually configure DNS server address on this interface to `10.1.1.1` and `10.2.2.2`. Then launched DSC again.
```
PS C:\DscTest2> Start-DscConfiguration -UseExisting -Verbose -Wait
VERBOSE: Perform operation 'Invoke CimMethod' with following parameters, ''methodName' = ApplyConfiguration,'className' = MSFT_DSCLocalConfigurationManager,'namespaceName' =
root/Microsoft/Windows/DesiredStateConfiguration'.
VERBOSE: An LCM method call arrived from computer INTERNETVM2 with user sid S-1-5-21-2186093478-4284278558-843089667-500.
VERBOSE: [INTERNETVM2]:                            [] Starting consistency engine.
VERBOSE: [INTERNETVM2]:                            [] Checking consistency for current configuration.
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Resource ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Test     ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]
VERBOSE: [INTERNETVM2]:                            [[NetIPInterface]Ethernet_2_IPv4_Dhcp] Test-TargetResource: Checking settings on net IP interface for alias 'Ethernet 2' and address
family 'IPv4'.
VERBOSE: [INTERNETVM2]:                            [[NetIPInterface]Ethernet_2_IPv4_Dhcp] Get-NetworkIPInterface: Getting settings for net IP interface for alias 'Ethernet 2' and address
family 'IPv4'.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Test     ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]  in 0.9340 seconds.
VERBOSE: [INTERNETVM2]: LCM:  [ Skip   Set      ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]
VERBOSE: [INTERNETVM2]: LCM:  [ End    Resource ]  [[NetIPInterface]Ethernet_2_IPv4_Dhcp]
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Resource ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Test     ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Test-TargetResource: Checking the DNS server addresses.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Get-DnsClientServerStaticAddress: Getting staticly assigned DNS server IPv4 address
for interface alias 'Ethernet 2'.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Get-DnsClientServerStaticAddress: Statically assigned DNS server IPv4 address for
interface alias 'Ethernet 2' is '10.1.1.1,10.2.2.2'.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Test-TargetResource: DNS server addresses are not correct. Expected "", actual
"10.1.1.1,10.2.2.2".
VERBOSE: [INTERNETVM2]: LCM:  [ End    Test     ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]  in 1.1830 seconds.
VERBOSE: [INTERNETVM2]: LCM:  [ Start  Set      ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Set-TargetResource: Applying the DNS server addresses.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Get-DnsClientServerStaticAddress: Getting staticly assigned DNS server IPv4 address
for interface alias 'Ethernet 2'.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Get-DnsClientServerStaticAddress: Statically assigned DNS server IPv4 address for
interface alias 'Ethernet 2' is '10.1.1.1,10.2.2.2'.
VERBOSE: [INTERNETVM2]:                            [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress] Set-TargetResource: DNS server addresses were set to the desired state of DHCP.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Set      ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]  in 0.8300 seconds.
VERBOSE: [INTERNETVM2]: LCM:  [ End    Resource ]  [[DnsServerAddress]Ethernet_2_IPv4_DnsServerAddress]
VERBOSE: [INTERNETVM2]:                            [] Consistency check completed.
VERBOSE: Operation 'Invoke CimMethod' complete.
VERBOSE: Time taken for configuration job to complete is 3.752 seconds
```
Again, the verbose messages confirm the interface was configured to obtain DNS server addresses via DHCP. Manually inspecting the setting via the UI confirms this.

## Testing Environment ##
#### The operating system the target node is running
```
PS C:\PowerShell> Get-ComputerInfo -Property @( 'OsName','OsOperatingSystemSKU','OSArchitecture','WindowsVersion','WindowsBuildLabEx','OsLanguage','OsMuiLanguages')


OsName               : Microsoft Windows Server 2019 Datacenter Evaluation
OsOperatingSystemSKU : 80
OsArchitecture       : 64-bit
WindowsVersion       : 1809
WindowsBuildLabEx    : 17763.1.amd64fre.rs5_release.180914-1434
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

```
#### Version and build of PowerShell the target node is running
```
PS C:\PowerShell> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.17763.316
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17763.316
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```
